File: flutter_zoom_videosdk-2.4.0/android/src/main/java/com/flutterzoom/videosdk/FlutterZoomVideoSdkAudioHelper.java
package com.flutterzoom.videosdk;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.flutterzoom.videosdk.convert.FlutterZoomVideoSdkErrors;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import us.zoom.sdk.ZoomVideoSDK;
import us.zoom.sdk.ZoomVideoSDKAudioHelper;
import us.zoom.sdk.ZoomVideoSDKErrors;
import us.zoom.sdk.ZoomVideoSDKUser;

public class FlutterZoomVideoSdkAudioHelper {

    private Activity activity;
    private Map<String, ZoomVideoSDKAudioHelper.ZoomVideoSDKAudioDevice> audioDeviceMap;

    FlutterZoomVideoSdkAudioHelper(Activity activity) {
        this.activity = activity;
    }

    private ZoomVideoSDKAudioHelper getAudioHelper() {
        ZoomVideoSDKAudioHelper audioHelper = null;
        try {
            audioHelper = ZoomVideoSDK.getInstance().getAudioHelper();
            if (audioHelper == null) {
                throw new Exception("No Audio Helper Found");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return audioHelper;
    }

    public void canSwitchSpeaker(@NonNull MethodChannel.Result result) {
        result.success(getAudioHelper().canSwitchSpeaker());
    }

    public void getSpeakerStatus(@NonNull MethodChannel.Result result) {
        result.success(getAudioHelper().getSpeakerStatus());
    }

    public void muteAudio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String userId = (String) args.get("userId");
        if (userId == null || userId.isEmpty()) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "userId is null or empty", null);
        }
        ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
        if (user != null) {
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().muteAudio(user)));
                }
            });
        } else {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "User not found", null);
        }
    }

    public void unmuteAudio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String userId = (String) args.get("userId");
        if (userId == null || userId.isEmpty()) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "userId is null or empty", null);
        }
        ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
        if (user != null) {
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().unMuteAudio(user)));
                }
            });
        } else {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "User not found", null);
        }
    }

    public void muteAllAudio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        boolean allowUnmute = (Boolean) args.get("allowUnmute");
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().muteAllAudio(allowUnmute)));
            }
        });
    }

    public void unmuteAllAudio(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().unmuteAllAudio()));
            }
        });
    }

    public void allowAudioUnmutedBySelf(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        boolean allowUnmute = (Boolean) args.get("allowUnmute");
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().allowAudioUnmutedBySelf(allowUnmute)));
            }
        });
    }

    public void setSpeaker(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        boolean isOn = (Boolean) args.get("isOn");

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().setSpeaker(isOn)));
            }
        });
    }

    public void startAudio(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().startAudio()));
            }
        });
    }

    public void stopAudio(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().stopAudio()));
            }
        });
    }

    public void subscribe(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().subscribe()));
            }
        });
    }

    public void unsubscribe(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().unSubscribe()));
            }
        });
    }

    public void resetAudioSession(@NonNull MethodChannel.Result result) {
        result.success(false);
    }

    public void cleanAudioSession(@NonNull MethodChannel.Result result) {
        result.success(null);
    }

    public void getAudioDeviceList(@NonNull MethodChannel.Result result) {
        List<ZoomVideoSDKAudioHelper.ZoomVideoSDKAudioDevice> audioDeviceList = getAudioHelper().getAudioDeviceList();
        for (ZoomVideoSDKAudioHelper.ZoomVideoSDKAudioDevice device : audioDeviceList) {
            if (audioDeviceMap == null) {
                audioDeviceMap = new java.util.HashMap<>();
            }
            audioDeviceMap.put(device.getAudioName(), device);
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkAudioDevice.jsonAudioDeviceArray(audioDeviceList));
            }
        });
    }

    public void getUsingAudioDevice(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkAudioDevice.jsonAudioDevice(getAudioHelper().getUsingAudioDevice()));
            }
        });
    }

    public void switchToAudioSourceType(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String deviceName = (String) args.get("deviceName");
        ZoomVideoSDKAudioHelper.ZoomVideoSDKAudioDevice device = audioDeviceMap.get(deviceName);
        if (device == null) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "Audio device not found", null);
            return;
        }
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getAudioHelper().switchToAudioSourceType(device)));
            }
        });
    }

}

File: flutter_zoom_videosdk-2.4.0/ios/Classes/FlutterZoomVideoSdkAudioHelper.m
#import "FlutterZoomVideoSdkAudioHelper.h"
#import "FlutterZoomVideoSdkUser.h"
#import "FlutterZoomVideoSdkAudioDevice.h"
#import "JSONConvert.h"
#import <AVFAudio/AVAudioSession.h>

@implementation FlutterZoomVideoSdkAudioHelper

- (ZoomVideoSDKAudioHelper *)getAudioHelper {
    ZoomVideoSDKAudioHelper* audioHelper = nil;
    @try {
        audioHelper = [[ZoomVideoSDK shareInstance] getAudioHelper];
        if (audioHelper == nil) {
            NSException *e = [NSException exceptionWithName:@"NoAudioHelperFound" reason:@"No Audio Helper Found" userInfo:nil];
            @throw e;
        }
    } @catch (NSException *e) {
        NSLog(@"%@ - %@", e.name, e.reason);
    }
    return audioHelper;
}

-(void) canSwitchSpeaker: (FlutterResult) result {
    // There's no API for retrieving available outputs other than external devices
    // since iOS devices always have "Speaker".
    result(@YES);
}

-(void) getSpeakerStatus: (FlutterResult) result {
    for (AVAudioSessionPortDescription *output in [[[AVAudioSession sharedInstance] currentRoute] outputs]) {
        if ([[output portType] isEqualToString:@"Speaker"]) {
            result(@YES);
        } else {
            result(@NO);
        }
    }
}

-(void) setSpeaker:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    NSError *error;
    AVAudioSessionPortOverride override =  [call.arguments[@"isOn"] boolValue] ? AVAudioSessionPortOverrideSpeaker : AVAudioSessionPortOverrideNone;
    [[AVAudioSession sharedInstance] overrideOutputAudioPort:override error:&error];
    [[AVAudioSession sharedInstance] setActive:YES error:nil];
    if (error) {
        NSLog(@"%@", error);
    } else {
        result(@"ZoomVideoSDKError_Success");
    }
}

-(void) startAudio: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] startAudio])]);
    });
}

-(void) stopAudio: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] stopAudio])]);
    });
}

-(void) muteAudio:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString* userId = call.arguments[@"userId"];
        if ([userId length] == 0) {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"UserId is empty"
                                       details:nil]);
        }
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser:userId];
        if (user != nil) {
            result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] muteAudio:user])]);
        } else {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"User not found"
                                       details:nil]);
        }
    });
}

-(void) unMuteAudio:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString* userId = call.arguments[@"userId"];
        if ([userId length] == 0) {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"UserId is empty"
                                       details:nil]);
        }
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser:userId];
        if (user != nil) {
            result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] unmuteAudio:user])]);
        } else {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"User not found"
                                       details:nil]);
        }
    });
}

-(void) muteAllAudio:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] muteAllAudio:[call.arguments[@"allowUnmute"] boolValue]])]);
    });
}

-(void) unmuteAllAudio: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] unmuteAllAudio])]);
    });
}

-(void) allowAudioUnmutedBySelf:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] allowAudioUnmutedBySelf:[call.arguments[@"allowUnmute"] boolValue]])]);
    });
}

-(void) subscribe: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] subscribe])]);
    });
}

-(void) unSubscribe: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getAudioHelper] unSubscribe])]);
    });
}

-(void) resetAudioSession: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([[self getAudioHelper] resetAudioSession]) {
            result(@YES);
        } else {
            result(@NO);
        }
    }
  );
}

-(void) cleanAudioSession: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        [[self getAudioHelper] cleanAudioSession];
    });
}

-(void) getAvailableAudioOutputRoute: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSArray <ZoomVideoSDKAudioDevice*>* audioDeviceList = [[self getAudioHelper] getAvailableAudioOutputRoute];
        if (self->_outputAudioDeviceMap == nil) {
            self->_outputAudioDeviceMap = [NSMutableDictionary dictionary];
        }
        [self->_outputAudioDeviceMap removeAllObjects];

        for (ZoomVideoSDKAudioDevice *device in audioDeviceList) {
            if (device && [device getAudioName]) {
                NSString* deviceName = [device getAudioName];
                self->_outputAudioDeviceMap[deviceName] = device;
            }
        }
        result([FlutterZoomVideoSdkAudioDevice mapAudioDeviceArray: audioDeviceList]);
    });
}

-(void) setAudioOutputRoute:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    NSString* deviceName = call.arguments[@"deviceName"];
    ZoomVideoSDKAudioDevice *device = _outputAudioDeviceMap[deviceName];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        result(@([[self getAudioHelper] setAudioOutputRoute: device]));
    });
}

-(void) getCurrentAudioOutputRoute: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([FlutterZoomVideoSdkAudioDevice mapAudioDevice: [[self getAudioHelper] getCurrentAudioOutputRoute]]);
    });
}

-(void) getAvailableAudioInputsDevice: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSArray <ZoomVideoSDKAudioDevice*>* audioDeviceList = [[self getAudioHelper] getAvailableAudioInputsDevice];
        if (self->_inputAudioDeviceMap == nil) {
            self->_inputAudioDeviceMap = [NSMutableDictionary dictionary];
        }
        [self->_inputAudioDeviceMap removeAllObjects];

        for (ZoomVideoSDKAudioDevice *device in audioDeviceList) {
            if (device && [device getAudioName]) {
                NSString* deviceName = [device getAudioName];
                self->_inputAudioDeviceMap[deviceName] = device;
            }
        }
        result([FlutterZoomVideoSdkAudioDevice mapAudioDeviceArray: audioDeviceList]);
    });
}

-(void) setAudioInputDevice:(FlutterMethodCall *)call withResult:(FlutterResult) result {
    NSString* deviceName = call.arguments[@"deviceName"];
    ZoomVideoSDKAudioDevice *device = _inputAudioDeviceMap[deviceName];
    dispatch_async(dispatch_get_main_queue(), ^{
        result(@([[self getAudioHelper] setAudioInputDevice: device]));
    });
}

-(void) getCurrentAudioInputDevice: (FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([FlutterZoomVideoSdkAudioDevice mapAudioDevice: [[self getAudioHelper] getCurrentAudioInputDevice]]);
    });
}

@end
